### PR TITLE
Allow devlocal user to choose the type of user to create

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

tl;dr:
- Can create a new user of any type from the devlocal sign in page and redirects to the appropriate site
- Servernames are now collected passed through as a struct to auth handlers instead of individually
- Cypress tests now POST to the login form instead of searching for available button
- Devlocal Sign In page was styled a bit
- Adds many more tests for devlocal

---

It would be helpful if local sign in allowed you to create a new user of any type and sign you into the appropriate application.  This PR adds a drop down for the user type and then will create the appropriate associated models when it creates the user. Finally it will redirect to the appropriate site for that user type.

The logic that creates a user will look in the request for data including First Name, Last Name, Phone Number, and Email. If any of those are present it will use those instead of defaults. (These extra ones were not added to the html form). This allows a regular POST request to `/devlocal-auth/create` to create a user of any type instead of having to do DB operations.  In turn that provides a hook that means a Load Testing application can create these users at will.

Finally, I've limited the page to returning only 25 users.  Load testing has created 1000s of users and this page seems to take forever to load unless I have a limit.  I'm not convinced we're ready for pagination so either 25 is good or a *slightly* higher number. But we can't do unlimited anymore.

## Reviewer Notes

The change here got a little big because of the need to redirect to the correct application when a user was created of a type that is different than the application you start from.  If there's a way to make this PR more readable I'm happy to do it.

## Setup

Run the server and the client, go to create a new devlocal user and select the type from the dropdown.  Then submit the form to create a new user.

## Code Review Verification Steps

* [x] Request review from a member of a different team.

## Screenshots

Screen Shot 2019-03-26 at 1 53 40 PM

### CODE REVIEW

File: pkg/auth/authentication/devlocal.go [hunk 1]
```diff
@@ -77,11 +78,17 @@ func (h UserListHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+					<option value="office">Office</option>
+					<option value="tsp">TSP</option>
+					<option value="dps">DPS</option>
+				</select>
+				<button type="submit" data-hook="new-user-login">Login as a New User</button>
 			</p>
 		</form>
 	`))
-	err = t.Execute(w, identities)
+	err = t.Execute(w, identities[:25])
```
Comment (chrisgilmerproj) [inline pkg/auth/authentication/devlocal.go]: This may interrupt e2e tests. The e2e tests may need a better way to log in as opposed to finding the name in the list and selecting the button.

Comment (chrisgilmerproj) [reply pkg/auth/authentication/devlocal.go]: e2e tests now use a POST to login instead of looking for a button.

File: pkg/auth/authentication/devlocal.go [hunk 2]
```diff
@@ -58,30 +58,52 @@ func (h UserListHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+			  {{end}}
+			</div>
+
+			<div class="col-md-4">
+			  <h2 class="mt-4">Create a New User</h1>
+			  <form method="post" action="/devlocal-auth/new">
+				<p>
+				  <input type="hidden" name="gorilla.csrf.Token" value="` + csrfToken + `">
+				  <select name="userType">
+					<option value="milmove">MilMove</option>
```
Comment (stangah) [inline pkg/auth/authentication/devlocal.go]: As a lazy user I think I'd rather have a list of buttons instead of a dropdown so I can do less clicking

**Comment by chrisgilmerproj** [comment]
> This looks pretty good to me, but how does this get us away from using the devlocal page to login for Cypress?

It doesn't get us away from devlocal signin. It gets us away from using the devlocal signin page. So effectively this would get rid of `cy.signInAsUser()` and replace it with `cy.signInAsUserPostRequest()`, which is a direct POST. No longer would we need to use the button for the user we want to interact with.  That should mean we don't need to worry about how cypress renders the devlocal signin page, and thus no need to modify the behavior of the cookies being generated for the forms there.

FWIW, this PR also limits the number of returned users, which makes it mostly unusable for automated testing.

File: pkg/auth/authentication/devlocal.go (hunk 2 above)
Comment (chrisgilmerproj) [reply pkg/auth/authentication/devlocal.go]: I'll probably come back and fix this at some point, good call.

**Comment by chrisgilmerproj** [comment]
Folks - I had a little trouble merging master in this morning and that added a few more bugs I had to fix. But this is Ready! Would really appreciate a review on this so I can get back to load testing.

**Comment by stangah** [comment]
I'm having trouble with being redirected to a logged in state after making a new user, seems to be the cookie issue you're dealing with? We talked about maybe POSTing to the respective hostname rather than using a relative path, hopefully that helps

**Comment by chrisgilmerproj** [comment]
I'm going to try to fix the redirect with the suggestion from @stangah . If not then I'll probably split that into a separate PR and pull the changes from this PR. I don't need it but it sure was a nice convenience.

**Comment by chrisgilmerproj** [comment]
So posting to the different subdomains doesn't work here because of CSRF protection. What I needed to do was set the Cookie Domain explicitly. I've done that but I'm afraid its a little odd how I've done it. I'll take another pass at this and try to clean it up.

**Comment by chrisgilmerproj** [comment]
New Screenshot:

Screen Shot 2019-04-05 at 4 29 07 PM

File: cypress/support/commands.js [hunk 3]
```diff
@@ -193,16 +249,14 @@ Cypress.Commands.add('logout', () => {
...
-  cy.clearCookies();
-  cy.setupBaseUrl(baseUrl);
+Cypress.Commands.add('setBaseUrlAndClearAllCookies', userType => {
+  [milmoveBaseURL, officeBaseURL, tspBaseURL].forEach(url => {
+    Cypress.config('baseUrl', url);
+    cy.visit('/');
+    cy.clearCookies();
+  });
+  const baseUrl = userTypeToBaseURL[userType]; // eslint-disable-line security/detect-object-injection
+  Cypress.config('baseUrl', baseUrl);
```
Comment (chrisgilmerproj) [inline cypress/support/commands.js]: @lynzt - in my branch I ended up rewriting this to take advantage of constants.

Comment (lynzt) [reply cypress/support/commands.js]: @chrisgilmerproj I like your `foreach` much better... I was going to go back and refactor, but you beat me to it :)
